profiler/data.py

```diff
@@ -0,0 +1,137 @@
+        cycles_elapsed: int,
+        peak_memory: int,
+        other_stats: str = None,
+    ):
+        self.process = process
+        self.rt = rt
+        self.ut = ut
+        self.st = st
+        self.max_set_size = max_set_size
+        self.page_reclaims = page_reclaims
+        self.page_faults = page_faults
+        self.peak_memory = peak_memory
+        self.cycles_elapsed = cycles_elapsed
+        self.process = process
+        self.p1out = p1out
+        self.p2out = p2out
+        self.other_stats = other_stats
+
+    def __repr__(self):
+        return (
+            f"{self.__class__} process {self.process} rt {self.rt} ut {self.ut} st {self.st} "
+            f"max mem {self.max_set_size} p1out {self.p1out} p2out {self.p2out} other_stats {self.other_stats} "
+            f"somacore_v {self.somacore_version} tiledbsoma_v {self.tiledbsoma_version}"
+        )
+
+
+def extract_key_from_filename(filename):
+    match = re.match("(.+)\.run", filename)
+    assert match
+    return match.groups()[0]
+
+
+class ProfilerDict:
+    """Helper Dictionary class for implementing runs DB using file system.
+    Any class that implements a persistent dictionary can be used as the
+    main backing of the runs data"""
+
+    def __init__(self, path):
+        self.path = path
+
+    def items(self):
+        result = []
+        dir_list = os.listdir(self.path)
+        for filename in dir_list:
+            file = open(f"{self.path}/{filename}", "rb")
+            data = pickle.load(file)
+            file.close()
+            result.append([extract_key_from_filename(filename), data])
+        return result
+
+    def __iter__(self):
+        for item in self.items():
+            yield item
+
+    def __setitem__(self, key, value):
+        filename = f"{self.path}/{key}.run"
+        with open(filename, "wb") as f:
+            pickle.dump(value, f)
+
+    def __getitem__(self, key):
+        filename = f"{self.path}/{key}.run"
+        file = open(filename, "rb")
+        data = pickle.load(file)
+        file.close()
+        return data
+
+    def __len__(self):
+        dir_list = os.listdir(self.path)
+        return len(dir_list)
+
+    def close(self):
+        pass
+
+
+PROFILE_PATH = "./profile_path"
+
+
+class ProfileDB:
+    """Represents runs database relying on a persistent dictionary
+    Options are filesystem, SQLiteDict, TileDB and DynamoDB"""
+
+    if not os.path.exists(PROFILE_PATH):
+        os.mkdir(PROFILE_PATH)
+    profile_dict = ProfilerDict(PROFILE_PATH)
+
+    def print(self):
+        print("Printing DB")
+        print(f"db size = {len(self.profile_dict)}")
+        for key, value in self.profile_dict.items():
+            print(f"db {key} = {value}")
+
+    def addProfileData(self, data: ProfileData):
+        self.profile_dict[f"{data.now}"] = data
+
+    def extractProfileData(self, process: str, metric: str):
+        result = {}
+        for key, value in self.profile_dict.items():
```

Comment:
Since we want to find profiler runs by the "process" value, the db should use the "process" value as the key, rather than a timestamp. Then it can find all runs for a given process efficiently without iterating through (worst case) all of the runs.

In fact, it seems like the file-based DB implementation could use a dir structure with process subdirs as the first "key", and timestamps file names as the second "key".